File: wordtype.cpp
/*********************************************************************
 *
 *  wordtype.cpp
 *
 *  Implementation of wordtype.h
 *
 ********************************************************************/

#include <iostream>
#include <string>
#include "wordtype.h"

Word :: Word(const std::string s) : s(s) {}

FeatureWord :: FeatureWord(const std::string s) : s(s), f("") {}
FeatureWord :: FeatureWord(const std::string s, const std::string f) : s(s)
{ FeatureWord::setFeatures(f); }

A :: A(const std::string s) : Word(s) {}

Adv :: Adv(const std::string s) : Word(s) {}

C :: C(const std::string s) : FeatureWord(s) {}
C :: C(const std::string s, const std::string f) : FeatureWord(s, f) {}
C :: void setFeatures(const std::string f)
{
  try
  {
    switch (f)
    {
    case "[+Q]":
    case "[+Q][+wh]":
    case "[+wh]":
    case "":
      this->f = f;
      break;
    default:
      throw (f);
    }
  }
  catch (const std::string s)
    std::cerr << "Invalid feature string for complementizer: " << s <<
      std::endl;
}

D :: D(const std::string s) : FeatureWord(s) {}
D :: D(const std::string s, const std::string f) : FeatureWord(s, f) {}
D :: void setFeatures(const std::string f)
{
  try
  {
    if (f == "[+wh]")
      this->f = f;
    else
      throw (f);
  }
  catch (const std::string s)
    std::cerr << "Invalid feature string for determiner: " << s <<
      std::endl;
}

N :: N(const std::string s) : Word(s) {}

Neg :: Neg(const std::string s) : Word(s) {}

P :: P(const std::string s) : Word(s) {}

T :: T(const std::string s) : FeatureWord(s) {}
T :: T(const std::string s, const std::string f) : FeatureWord(s, f) {}
T :: void setFeatures(const std::string f)
{
  try
  {
    switch (f)
    {
    case "[-tense]":
    case "[+past]":
    case "[+future]":
      this->f = f;
      break;
    default:
      throw (f);
    }
  }
  catch (const std::string s)
    std::cerr << "Invalid feature string for tense word: " << s <<
      std::endl;
}

V :: V(const std::string s) : Word(s) {}

File: wordtype.py
# wordtype.py
#
# Define parts of speech possible for a word.

class Word:
    _s = None
    
    def __init__(self, s):
        self._s = s

    def getWord(self):
        return self._s

    def getType(self):
        return self.__class__.__name__

class FeatureWord(Word):
    _f = None

    def __init__(self, s, f):
        self._s = s
        self._f = f
    
    def getFeatures(self):
        return self._f

    def setFeatures(self, f):
        raise NotImplementedError()

    class InvalidFeature(Exception):
        pass

class A(Word):
    pass

class Adv(Word):
    pass

class C(FeatureWord):
    def setFeatures(self, f):
        try:
            if f in ["[+Q]", "[+Q][+wh]", "[+wh]", "[+TOP]", None]:
                self._f = f
        except:
            raise self.InvalidFeature("Invalid feature string for "
                                      "complementizer: " + f)
        
class D(FeatureWord):
    def setFeatures(self, f):
        try:
            if f in ["[+wh]", None]:
                self._f = f
        except:
            raise self.InvalidFeature("Invalid feature string for"
                                      "determiner: " + f)

class N(Word):
    pass

class Neg(Word):
    pass

class P(Word):
    pass

class T(FeatureWord):
    def setFeatures(self, f):
        try:
            if f in ["[-tense]", "[+past]", "[+future]", None]:
                self._f = f
        except:
            raise self.InvalidFeature("Invalid feature string for "
                                      "tense word: " + f)

class V(Word):
    pass
